content/blog/2022-05-20-Banked_Memory_Compiler_Backend_in_MLIR/index.md

```diff
@@ -0,0 +1,302 @@
+
+Calyx and the Calyx dialect also need some modifications to support a wider range of memories that can be generated by AMC. For example, Calyx does not currently support multiported memories or memories that generate verilog with a specific memory primtive (BRAM, UltraRAM, etc.). In theory, these primtives could be added through a Calyx library, but this is also currently unsupported by the Calyx MLIR Dialect.
+
+As a result, the lowering from AMC to Calyx only support the amc.alloc and amc.create_port operations. All ports also must be single latency, read/write, and latency-sensitive. Here is an example lowering from AMC to Calyx:
+```mlir
+// Input to AMC to Calyx lowering
+amc.memory @test1(!amc.port<512xf32, rw, 1>, !amc.port<512xf32, rw, 1>) {
+  %0 = amc.alloc : !amc.memref<512xf32>
+  %1 = amc.alloc : !amc.memref<512xf32>
+  %2 = amc.create_port(%0 : !amc.memref<512xf32>) : !amc.port<512xf32, rw, 1>
+  %3 = amc.create_port(%1 : !amc.memref<512xf32>) : !amc.port<512xf32, rw, 1>
+  amc.extern %2, %3 : !amc.port<512xf32, rw, 1>, !amc.port<512xf32, rw, 1>
+}
+
+// Output
+calyx.program "test1" {
+  calyx.component @test1(%port_0.addr0: i9, %port_0.write_data: i32, %port_0.write_en: i1, %port_1.addr0: i9, %port_1.write_data: i32, %port_1.write_en: i1, %clk: i1 {clk}, %reset: i1 {reset}, %go: i1 {go}) -> (%port_0.read_data: i32, %port_1.read_data: i32, %done: i1 {done}) {
+    %bank0.addr0, %bank0.write_data, %bank0.write_en, %bank0.clk, %bank0.read_data, %bank0.done = calyx.memory @bank0 <[512] x 32> [9] : i9, i32, i1, i1, i32, i1
+    %bank1.addr0, %bank1.write_data, %bank1.write_en, %bank1.clk, %bank1.read_data, %bank1.done = calyx.memory @bank1 <[512] x 32> [9] : i9, i32, i1, i1, i32, i1
+    calyx.wires {
+      calyx.assign %bank0.clk = %clk : i1
+      calyx.assign %bank1.clk = %clk : i1
+      calyx.assign %bank0.write_data = %port_0.write_data : i32
+      calyx.assign %bank0.write_en = %port_0.write_en : i1
+      calyx.assign %port_0.read_data = %bank0.read_data : i32
+      calyx.assign %bank0.addr0 = %port_0.addr0 : i9
+      calyx.assign %bank1.write_data = %port_1.write_data : i32
+      calyx.assign %bank1.write_en = %port_1.write_en : i1
+      calyx.assign %port_1.read_data = %bank1.read_data : i32
+      calyx.assign %bank1.addr0 = %port_1.addr0 : i9
+    }
+    calyx.control {
+    }
+  }
+}
+```
```

Comment:
Is it worth mentioning whether it's possible to go the rest of the way to RTL from here?